File: NewsFeed/app/src/main/java/com/example/newsfeed/NewsListAdapter.java
package com.example.newsfeed;

import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;


import java.util.List;

public class NewsListAdapter extends RecyclerView.Adapter<NewsListAdapter.NewsViewHolder> {
    private List<News> newsList;
    private @Nullable ItemClickListener listener;

    public NewsListAdapter(List<News> newsList, @Nullable ItemClickListener listener) {
        this.newsList = newsList;
        this.listener = listener;
    }

    @NonNull
    @Override
    public NewsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(
                parent.getContext())
                .inflate(
                        R.layout.item_news,
                        null,
                        false);
        RecyclerView.LayoutParams params = new RecyclerView.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        );
        view.setLayoutParams(params);
        return new NewsViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull NewsViewHolder holder, final int position) {
        final News news = newsList.get(position);
        holder.profilePhoto.setImageResource(news.getProfilePhoto());
        holder.postImage.setImageResource(news.getPostImage());
        holder.author.setText(news.getAuthor());
        String s = "<b>"+ news.getAuthor()+ "</b>" + " "+ news.getPostText();
        holder.postText.setText(Html.fromHtml(s));
        holder.date.setText(news.getDate());
        holder.likesCnt.setText(news.getLikesCnt()+" likes");
        holder.isLiked = news.isLiked();
        holder.isSaved = news.isSaved();

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (listener!=null)
                    listener.ItemClick(position, news);
            }
        });

    }

    @Override
    public int getItemCount() {
        return newsList.size();
    }

    public class NewsViewHolder extends RecyclerView.ViewHolder{
        TextView author;
        TextView likesCnt;
        TextView postText;
        TextView date;
        ImageView postImage;
        ImageView profilePhoto;
        ImageView likeBtn;
        ImageView saveBtn;
        boolean isLiked;
        boolean isSaved;

        public NewsViewHolder(@NonNull final View itemView){
            super(itemView);
            author = itemView.findViewById(R.id.author);
            likesCnt = itemView.findViewById(R.id.likesCnt);
            postImage = itemView.findViewById(R.id.postImage);
            postText = itemView.findViewById(R.id.postText);
            date = itemView.findViewById(R.id.date);
            profilePhoto = itemView.findViewById(R.id.profilePhoto);
            likeBtn = itemView.findViewById(R.id.likeBtn);
            saveBtn = itemView.findViewById(R.id.saveBtn);

            likeBtn.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (!isLiked){
                        likeBtn.setImageResource(R.drawable.liked);
                        isLiked = true;
                    } else {
                        likeBtn.setImageResource(R.drawable.like);
                        isLiked = false;
                    }
                }
            });
            saveBtn.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (!isSaved){
                        saveBtn.setImageResource(R.drawable.saved);
                        isSaved = true;
                    } else {
                        saveBtn.setImageResource(R.drawable.save);
                    }
                }
            });
        }
    }

    interface ItemClickListener{
        void ItemClick(int position, News item);
    }
}

File: Hw4/app/src/main/java/com/example/hw4/NewsListAdapter.kt
package com.example.hw4

import android.annotation.SuppressLint
import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import java.lang.Exception

class NewsListAdapter(private var newsList: List<News>, private var listener: ItemClickListener):
    RecyclerView.Adapter<NewsListAdapter.NewsViewHolder>() {

    interface ItemClickListener {
        fun itemClick(position: Int, item: News)
    }

    @SuppressLint("InflateParams")
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NewsViewHolder {
        val view = LayoutInflater.from(
            parent.context
        )
            .inflate(
                R.layout.item_news,
                null,
                false
            )
        val params = RecyclerView.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        view.layoutParams = params
        return NewsViewHolder(view)
    }

    override fun getItemCount(): Int {
        return newsList.size
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: NewsViewHolder, position: Int) {
        val news = newsList[position]
        holder.profilePhoto.setImageResource(news.profilePhoto)
        holder.postImage.setImageResource(news.postImage)
        holder.author.text = news.author
        val s = "<b>" + news.author + "</b>" + " " + news.postText
        holder.postText.text = Html.fromHtml(s)
        holder.date.text = news.date
        holder.likesCnt.text = news.likesCnt.toString() + " likes"
        holder.isLiked = news.isLiked
        holder.isSaved = news.isSaved

        holder.itemView.setOnClickListener {
            if (listener!=null) {
                listener.itemClick(position, news)
            }
        }
    }

    class NewsViewHolder(itemView: View) : ViewHolder(itemView) {
        var author: TextView = itemView.findViewById(R.id.author)
        var likesCnt: TextView = itemView.findViewById(R.id.likesCnt)
        var postText: TextView = itemView.findViewById(R.id.postText)
        var date: TextView = itemView.findViewById(R.id.date)
        var postImage: ImageView = itemView.findViewById(R.id.postImage)
        var profilePhoto: ImageView = itemView.findViewById(R.id.profilePhoto)
        private var likeBtn: ImageView = itemView.findViewById(R.id.likeBtn)
        private var saveBtn: ImageView = itemView.findViewById(R.id.saveBtn)
        var isLiked = false
        var isSaved = false

        init {
            likeBtn.setOnClickListener {
                isLiked = if (!isLiked) {
                    likeBtn.setImageResource(R.drawable.liked)
                    true
                } else {
                    likeBtn.setImageResource(R.drawable.like)
                    false
                }
            }
            saveBtn.setOnClickListener {
                if (!isSaved) {
                    saveBtn.setImageResource(R.drawable.saved)
                    isSaved = true
                } else {
                    saveBtn.setImageResource(R.drawable.save)
                }
            }
        }
    }
}